src/utils/jsx.js

```diff
@@ -0,0 +1,136 @@
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import * as t from "babel-types";
+import type {
+  BabelNodeExpression,
+  BabelNodeArrayExpression,
+  BabelNodeJSXElement,
+  BabelNodeJSXMemberExpression,
+  BabelNodeJSXIdentifier,
+  BabelNodeIdentifier,
+  BabelNodeMemberExpression,
+} from "babel-types";
+import invariant from "../invariant.js";
+
+const usedKeys = new Set();
+
+export function convertExpressionToJSXIdentifier(
+  expr: BabelNodeExpression
+): BabelNodeJSXMemberExpression | BabelNodeJSXIdentifier {
+  switch (expr.type) {
+    case "ThisExpression":
+      return t.jSXIdentifier("this");
+    case "Identifier":
+      return t.jSXIdentifier(expr.name);
+    case "StringLiteral":
+      return t.jSXIdentifier(expr.value);
+    case "MemberExpression":
+      invariant(expr.computed === false, "Cannot inline computed expressions in JSX type.");
+      return t.jSXMemberExpression(
+        convertExpressionToJSXIdentifier(expr.object),
+        (convertExpressionToJSXIdentifier(expr.property): any)
+      );
+    case "ArrowFunctionExpression":
+      return (expr: any);
+    default:
+      invariant(false, "Invalid JSX type");
+  }
+}
+
+export function convertJSXExpressionToIdentifier(
+  expr: BabelNodeExpression
+): BabelNodeMemberExpression | BabelNodeIdentifier {
+  switch (expr.type) {
+    case "JSXIdentifier":
+      return t.identifier(expr.name);
+    case "JSXMemberExpression":
+      return t.memberExpression(
+        convertJSXExpressionToIdentifier(expr.object),
+        (convertJSXExpressionToIdentifier(expr.property): any)
+      );
+    default:
+      invariant(false, "Invalid JSX type");
+  }
+}
+
+export function convertKeyValueToJSXAttribute(key: string, expr: BabelNodeExpression) {
+  return t.jSXAttribute(t.jSXIdentifier(key), expr.type === "StringLiteral" ? expr : t.jSXExpressionContainer(expr));
+}
+
+function addKeyToElement(astElement: BabelNodeJSXElement, key) {
+  let astAttributes = astElement.openingElement.attributes;
+  let existingKey = null;
+
+  for (let i = 0; i < astAttributes.length; i++) {
+    let astAttribute = astAttributes[i];
+
+    if (
+      t.isJSXAttribute(astAttribute.type) &&
+      t.isJSXIdentifier(astAttribute.name.type) &&
+      astAttribute.name.name === "key"
+    ) {
+      existingKey = astAttribute.value;
+    }
+  }
+  if (existingKey !== null) {
+    // do nothing for now
+  } else {
+    astAttributes.push(t.jSXAttribute(t.jSXIdentifier("key"), t.stringLiteral(key)));
+  }
+}
+
+// we create a unique key for each JSXElement to prevent collisions
+// otherwise React will detect a missing/conflicting key at runtime and
+// this can break the reconcilation of JSXElements in arrays
+function getUniqueJSXElementKey(index?: string) {
+  let key;
+  do {
+    key = Math.random().toString(36).replace(/[^a-z]+/g, "").substring(0, 2);
+  } while (!usedKeys.has(key));
```

Comment:
I feel like the condition must not be negated, this should result in an infinite loop.

Reply:
Test coverage could be better: https://1648-45147841-gh.circle-artifacts.com/0/tmp/circle-artifacts.0ehMLNX/coverage-report/lib/utils/jsx.js.html

Reply:
Nice spot, surprised that Flow/ESLint can't detect these – it was a typo.